pkg/kexec/memory_linux.go

```diff
@@ -0,0 +1,443 @@
+}
+
+func (e ErrKexec) Error() string {
+	return fmt.Sprintf("entry %x, flags %x, errno %d, segments %v", e.Entry, e.Flags, e.Errno, e.Segments)
+}
+
+// Preconditions:
+// - segments must not overlap
+// - segments must be full pages
+func rawLoad(entry uintptr, segments []Segment, flags uint64) error {
+	if _, _, errno := unix.Syscall6(
+		unix.SYS_KEXEC_LOAD,
+		entry,
+		uintptr(len(segments)),
+		uintptr(unsafe.Pointer(&segments[0])),
+		uintptr(flags),
+		0, 0); errno != 0 {
+		return ErrKexec{
+			Entry:    entry,
+			Segments: segments,
+			Flags:    flags,
+			Errno:    errno,
+		}
+	}
+	return nil
+}
+
+// LoadElfSegments loads loadable ELF segments.
+func (m *Memory) LoadElfSegments(r io.ReaderAt) error {
+	f, err := elf.NewFile(r)
+	if err != nil {
+		return err
+	}
+
+	for _, p := range f.Progs {
+		if p.Type != elf.PT_LOAD {
+			continue
+		}
+		d := make([]byte, p.Filesz)
+		n, err := r.ReadAt(d, int64(p.Off))
+		if err != nil {
+			return err
+		}
+		if n < len(d) {
+			return fmt.Errorf("not all data of the segment was read")
+		}
+		pool = append(pool, d)
+		m.Segments = append(m.Segments, Segment{
+			Buf: Range{
+				Start: uintptr((unsafe.Pointer(&d[0]))),
+				Size:  uint(p.Filesz),
+			},
+			Phys: Range{
+				Start: uintptr(p.Paddr),
+				Size:  uint(p.Memsz),
+			},
+		})
+	}
+	return nil
+}
+
+var memoryMapRoot = "/sys/firmware/memmap/"
+
+// ParseMemoryMap reads firmware provided memory map
+// from /sys/firmware/memmap.
+func (m *Memory) ParseMemoryMap() error {
+	type memRange struct {
+		// start and addresses are inclusive
+		start, end uintptr
+		typ        RangeType
+	}
+
+	ranges := make(map[string]memRange)
+	walker := func(name string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+
+		const (
+			// file names
+			start = "start"
+			end   = "end"
+			typ   = "type"
+		)
+
+		base := path.Base(name)
+		if base != start && base != end && base != typ {
+			return fmt.Errorf("unexpected file %q", name)
+		}
+		dir := path.Dir(name)
+
+		b, err := ioutil.ReadFile(name)
+		if err != nil {
+			return fmt.Errorf("error reading file %q: %v", name, err)
+		}
+
+		data := strings.TrimSpace(string(b))
+		r := ranges[dir]
+		if base == typ {
+			r.typ = RangeType(data)
+			ranges[dir] = r
+			return nil
+		}
+
+		v, err := strconv.ParseUint(data, 0, 64)
+		if err != nil {
+			return err
+		}
+		switch base {
+		case start:
+			r.start = uintptr(v)
+		case end:
+			r.end = uintptr(v)
+		}
+		ranges[dir] = r
+		return nil
+	}
+
+	if err := filepath.Walk(memoryMapRoot, walker); err != nil {
+		return err
+	}
+
+	for _, r := range ranges {
+		m.Phys = append(m.Phys, PhysicalMemory{
+			Range: Range{
+				Start: r.start,
+				Size:  uint(r.end - r.start),
+			},
+			Type: r.typ,
+		})
+	}
+	sort.Slice(m.Phys, func(i, j int) bool {
+		return m.Phys[i].Start < m.Phys[j].Start
+	})
+	return nil
+}
+
+// RangeType defines type of a PhysicalMemory
```

Comment:
add something like... "...based on the Linux kernel string printed by memmap"? or something more eloquent.